### PR TITLE
Add AssertJ migrations for jupiter assertions that contain messages.

### PR DESCRIPTION
**Pull request by bbogason**
Closes SEN-2506

Add 'doStaticImports: true' to all fixes

### CODE REVIEW

File: recipes/Java/AssertJ/From JUnit/JUnit_assertArrayEquals_message.yaml [hunk 1]
```diff
@@ -9,14 +9,33 @@ metadata:
...
+            methodcall:
+              type: org.junit.jupiter.api.Assertions
+      to: org.assertj.core.api.Assertions.assertThat({{{ arguments.1 }}}).as({{{ arguments.2 }}}).isEqualTo({{{ arguments.0 }}})
+  - rewrite:
+      applyIf:
+        markedElement:
+          is:
+            methodcall:
+              not:
+                type: org.junit.jupiter.api.Assertions
```
Comment (seanf) [inline recipes/Java/AssertJ/From JUnit/JUnit_assertArrayEquals_message.yaml:40]: A minor thing, but wouldn't it be a bit neater/cleaner to say `markedElement is methodcall type: org.junit.Assert` in these cases?

Ah. I guess it's because most recipes have a special case just for org.junit.jupiter, whereas org.junit and junit.framework have similar APIs.

**Review by seanf — APPROVED** [review]

File: recipes/Java/AssertJ/From JUnit/JUnit_assertArrayEquals.yaml [hunk 2]
```diff
@@ -12,10 +12,11 @@ search:
...
-      - type: org.junit.Assert
-      - type: org.junit.jupiter.api.Assertions
+    - type: org.junit.Assert
+    - type: org.junit.jupiter.api.Assertions
 availableFixes:
-  - name: Change to assertThat(actualArray).isEqualTo(expectedArray)
-    actions:
-      - rewrite:
-          to: org.assertj.core.api.Assertions.assertThat({{{ arguments.1 }}}).isEqualTo({{{ arguments.0 }}})
+- doStaticImports: true
```
[the remark this answers is not in the document]
Comment (seanf) [reply recipes/Java/AssertJ/From JUnit/JUnit_assertArrayEquals.yaml:18]: This problem also exists with 2021.2.3 and the released recipes.

[the remark this answers is not in the document]
Comment (seanf) [reply recipes/Java/AssertJ/From JUnit/JUnit_assertArrayEquals.yaml:18]: And with Sensei by Secure Code Warrior® 2021.2.10

[the remark this answers is not in the document]
Comment (bbogason) [reply recipes/Java/AssertJ/From JUnit/JUnit_assertArrayEquals.yaml:18]: I tested this with Sensei 2021.2.11-alpha.36 and it worked:

https://user-images.githubusercontent.com/89931719/140318756-70ce392f-562d-4836-ba37-81fd2660965b.mov

There was a bug that I fixed in SEN-2461, which has a fix-version 2021.2.11 so these recipes will work once that has been released. (And for older versions the recipes will still work, except they won't statically import `assertThat`)

File: recipes/Java/AssertJ/From JUnit/JUnit_assertArrayEquals_message.yaml (hunk 1 above)
Comment (bbogason) [reply recipes/Java/AssertJ/From JUnit/JUnit_assertArrayEquals_message.yaml:40]: Yes, for the majority of the rules, I want one behaviour for `org.junit.Assert`, `junit.framework.TestCase` and `junit.framework.Assert`, but a different behaviour for `org.junit.jupiter.api.Assertions` so I found this to be the cleanest solution (since there is no `else` clause for the `applyIf`).

File: recipes/Java/AssertJ/From JUnit/JUnit_assertArrayEquals.yaml (hunk 2 above)
[the remark this answers is not in the document]
Comment (bbogason) [reply recipes/Java/AssertJ/From JUnit/JUnit_assertArrayEquals.yaml:18]: But maybe this feature is flaky, I do sometimes see this being converted to `org.assertj.core.api.Assertions.assertFalse` instead of just `assertFalse`. 

I haven't figured out why it sometimes doesn't work...

[the remark this answers is not in the document]
Comment (seanf) [reply recipes/Java/AssertJ/From JUnit/JUnit_assertArrayEquals.yaml:18]: Sorry, I apparently got a bit confused about which version it's in, but I was reproducing this in a debugger, with pretty recent code (4c7bc7ab3470c1d2ab51b94eb26bc05fe1d6ff29 merged into SEN-2416-gray-out-disabled-recipes-in-recipe-editor).

com.sensei.intellij.JavaImportWriter#computeAndAddImports calls com.sensei.intellij.JavaImportWriter#computeImportToAddAndDequalify which is pretty hard to follow, but if `referencedClass` is resolved as `null`, no import will be added (static or otherwise). That's what is happening in my debug session.

I had a breakpoint on `if (!reference.isValid())`, with the condition `reference.toString().equals("PsiReferenceExpression:org.assertj.core.api.Assertions.assertThat")`

Not sure if it's relevant, but my repro involves editing the fix Yaml in the recipe editor and watching the fix preview.

[the remark this answers is not in the document]
Comment (seanf) [reply recipes/Java/AssertJ/From JUnit/JUnit_assertArrayEquals.yaml:18]: @bbogason 
The whole import seems to fail consistently (not even a class import) on the class JUnit5Assertions in sense-recipes. With v2021.2.11-alpha.38

[the remark this answers is not in the document]
Comment (seanf) [reply recipes/Java/AssertJ/From JUnit/JUnit_assertArrayEquals.yaml:18]: @bbogason 
Aha! I think it's because `JUnit5Assertions` already has a class called `Assertions` imported (the one from JUnit Jupiter). There's no actual name conflict for imports at the method level, but there is a conflict at the class level, which trips up the current implementation. This is not the case for `ReflectionCheckTests` in Sensei, thus there's no problem there.

I don't think these JUnit 5 recipes will use imports at all until we fix the class-conflict problem (because JUnit 5 tests will all use `org.junit.jupiter.api.Assertions`).

Also, I just realised that `JUnit5Assertions` has examples for JUnit 5 and JUnit 4, but not for 4 (`junit.framework.Assert`). (We'd be better off with one file for each version, because they have imports which conflict.)

[the remark this answers is not in the document]
Comment (seanf) [reply recipes/Java/AssertJ/From JUnit/JUnit_assertArrayEquals.yaml:18]: Come to think of it, JUnit 5 tests which use static imports as recommended: "The JUnit Team highly recommends that the methods defined in this class be used via _static imports_." should be okay. Just so long as they don't do this:

```java
import org.junit.jupiter.api.Assertions;
```

[the remark this answers is not in the document]
Comment (seanf) [reply recipes/Java/AssertJ/From JUnit/JUnit_assertArrayEquals.yaml:18]: https://securecodewarrior.atlassian.net/browse/SEN-2529
